src/cryptography/hazmat/primitives/smime.py

```diff
@@ -0,0 +1,109 @@
+# This file is dual licensed under the terms of the Apache License, Version
+# 2.0, and the BSD License. See the LICENSE file in the root of this repository
+# for complete details.
+
+from __future__ import absolute_import, division, print_function
+
+from enum import Enum
+
+from cryptography import x509
+from cryptography.hazmat.backends import _get_backend
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import ec, rsa
+from cryptography.utils import _check_byteslike
+
+
+class SMIMESignatureBuilder(object):
+    def __init__(self, data=None, signers=[]):
+        self._data = data
+        self._signers = signers
+
+    def set_data(self, data):
+        _check_byteslike("data", data)
+        if self._data is not None:
+            raise ValueError("data may only be set once")
```

Comment:
Data is mandatory and can only be set once, - maybe it would be better to move data initialization at class instantiation time? And class can be named simply `SMIME`:
```python
class SMIME(object):
    def __init__(self, data):
        _check_byteslike("data", data)
        self._data = data
        self._signers = []

    def add_signer(self, certificate, private_key, hash_algorithm):
        ...
        self._signers + [(certificate, private_key, hash_algorithm)]
        return self

    def sign(self, encoding, options, backend=None):
        ...
```

Reply:
This will be unnecessary within `sign` method:
```python
if self._data is None:
    raise ValueError("You must add data to sign")
```

Reply:
This class is following our Builder pattern (which we use in X509 quite a bit) in not having a public `__init__`. While passing `data` in the constructor is arguably more elegant, I'd rather be consistent with the rest of the API in this case. It also means if we have a consistent API path should we ever choose to support streaming contexts.

Reply:
Ok, I understand you.